File: src/Lexer.py
from src.Token import Token, Position
from src.Settings import LangData
from src.LexerState import LexerStateNormal

class Lexer:
  def __init__(self):
    self.reset()

  def reset(self):
    self.tokens = []
    self.currentToken = Token('', [0,0])
    self.state = LexerStateNormal(self)

  def tokenize(self, text):
    self.reset()
    lines = text.split('\n')
    for lineNo, line in enumerate(lines):
      line += '\n' # for comment end detection
      for charNo in range(len(line)):
        self.state.tokenize(line, Position(lineNo, charNo))

File: javascript/src/Lexer.js

class Lexer {
  constructor() {
    this.reset();
  }
  reset(parent) {
    this.tokens = [];
    this.currentToken = new Token('', new Position(0, 0));
    this.state = new LexerStateNormal(this);
  }
  tokenize(text) {
    this.reset();
    let lines = text.split('\n');
    let lineNo = 0;
    for (let line of lines) {
      line += '\n'; // for comment end detection
      for (let charNo = 0; charNo < line.length; charNo += 1) {
        this.state.tokenize(line, new Position(lineNo, charNo));
      }
      lineNo += 1;
    }
  }
}
